registry/storage/manifeststore.go

```diff
@@ -135,6 +147,47 @@ func (ms *manifestStore) verifyManifest(ctx context.Context, mnfst *manifest.Sig
 				errs = append(errs, distribution.ErrManifestBlobUnknown{Digest: fsLayer.BlobSum})
 			}
 		}
+
+		// image provides a local type for validating the image relationship.
+		type image struct {
+			Id     string
+			Parent string
+		}
+
+		// Process the history portion to ensure that the parent links are
+		// correctly represented. We serialize the image json, then walk the
+		// entries, checking the parent link.
+		var images []image
+		for _, entry := range mnfst.History {
+			var im image
+			if err := json.Unmarshal([]byte(entry.V1Compatibility), &im); err != nil {
+				errs = append(errs, err)
+			}
+
+			images = append(images, im)
+		}
+
+		// go back through each image, checking the parent link and rank
+		for i, im := range images {
```

Comment:
Alternative loop that is a little simpler (other than the reverse iteration maybe). Feel free to ignore.
```go
var parentID string
for i := len(images)-1; i > 0; i-- {
    if images[i].Parent != parentID {
        errs = append(errs, distribution.ErrManifestValidation{...}) // more generic error message?
    }
    parent = images[i]
}
```